Unset phone and visible-password keyboardType flags on Android

BrandonWilliamsCS opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes. 

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to React Native here: http://facebook.github.io/react-native/docs/contributing.html

Happy contributing!

-->

## Motivation

Attempted fix for issue #17248

## Test Plan

Added unit tests that cover the affected function and manually repeated the reproduction steps found in the issue.

Note: I had to apply the attached patch file to actually *run* the tests because they were not enabled. I didn't include this in the PR because it seemed like a secondary problem with possible ramifications (see comment near patched line) beyond this issue. For example, other, unrelated tests break when that line is uncommented.
[textInputTestEnable.patch.txt](https://github.com/facebook/react-native/files/1565807/textInputTestEnable.patch.txt)

If I should apply the patch to this PR or re-enable the tests in some other fashion, please let me know and I can do so.

## Related PRs

(none)

## Release Notes
<!--
Help reviewers and the release process by writing your own release notes

**INTERNAL and MINOR tagged notes will not be included in the next version's final release notes.**

  CATEGORY
[----------]        TYPE
[ CLI      ]   [-------------]      LOCATION
[ DOCS     ]   [ BREAKING    ]   [-------------]
[ GENERAL  ]   [ BUGFIX      ]   [-{Component}-]
[ INTERNAL ]   [ ENHANCEMENT ]   [ {File}      ]
[ IOS      ]   [ FEATURE     ]   [ {Directory} ]   |-----------|
[ ANDROID  ]   [ MINOR       ]   [ {Framework} ] - | {Message} |
[----------]   [-------------]   [-------------]   |-----------|

[CATEGORY] [TYPE] [LOCATION] - MESSAGE

 EXAMPLES:

 [IOS] [BREAKING] [FlatList] - Change a thing that breaks other things
 [ANDROID] [BUGFIX] [TextInput] - Did a thing to TextInput
 [CLI] [FEATURE] [local-cli/info/info.js] - CLI easier to do things with
 [DOCS] [BUGFIX] [GettingStarted.md] - Accidentally a thing/word
 [GENERAL] [ENHANCEMENT] [Yoga] - Added new yoga thing/position
 [INTERNAL] [FEATURE] [./scripts] - Added thing to script that nobody will see
-->

[ANDROID] [BUGFIX] [TextInput] - Fix an issue when swapping to and from the `'visible-password'` or `'phone-pad'` keyboard types.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -527,6 +527,13 @@ public void setAutoCapitalize(ReactEditText view, int autoCapitalize) {
 
   @ReactProp(name = "keyboardType")
   public void setKeyboardType(ReactEditText view, @Nullable String keyboardType) {
+    int passwordVisibilityFlag = InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD &
```
So this flag is only the visibility bit if I understand correctly? It won't unset the fact that it is a password or not.

BrandonWilliamsCS replied:
That is correct. We wouldn't want to set/unset general password status in `keyboardType`, only in `secureTextEntry`.

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -527,6 +527,13 @@ public void setAutoCapitalize(ReactEditText view, int autoCapitalize) {
+            passwordVisibilityFlag;
     int flagsToSet = InputType.TYPE_CLASS_TEXT;
     if (KEYBOARD_TYPE_NUMERIC.equalsIgnoreCase(keyboardType)) {
       flagsToSet = INPUT_TYPE_KEYBOARD_NUMBERED;
```
A bit later in this code we set the TYPE_TEXT_VARIATION_VISIBLE_PASSWORD flag, should we just set passwordVisibilityFlag instead?

BrandonWilliamsCS replied:
That's a good question. I think it comes down to how the field should behave when `keyboardType="visible-password"` but `secureTextEntry="false"`. Due to the way Android handles the flags, if we set the flag I called `passwordVisibilityFlag` without setting the `InputType.TYPE_TEXT_VARIATION_PASSWORD` flag, it will actually treat it like `InputType.TYPE_TEXT_VARIATION_URI`.

So three options I see:
1. Somehow forbid `keyboardType="visible-password"` with `secureTextEntry="false"`
2. Let `keyboardType="visible-password"` override the `secureTextEntry` prop (equivalent to the current PR code)
3. Allow the consumer to set the props inconsistently (probably with a warning) and risk an unexpected `InputType.TYPE_TEXT_VARIATION_URI`

I should probably leave that decision up to the core maintainers.

janicduplessis replied:
Makes sense, let's keep the current behavior (current PR code).

janicduplessis commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -527,6 +527,13 @@ public void setAutoCapitalize(ReactEditText view, int autoCapitalize) {
+    int passwordVisibilityFlag = InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD &
+        ~InputType.TYPE_TEXT_VARIATION_PASSWORD;
+
+    int keyboardTypeFlags = INPUT_TYPE_KEYBOARD_NUMBERED |
```
Let's move this and passwordVisibilityFlag to private static final members in the class.

BrandonWilliamsCS replied:
Will do, commit incoming.

janicduplessis commented:
Thanks for figuring this out, this flag code is pretty hard to understand -_-

BrandonWilliamsCS commented:
Glad I could help!
I moved the flag values to private static finals and added a comment noting the quirk, so that future contributions can have a little more context into the mess that is coordinating InputType flags between declarative and imperative contexts. Let me know if there's more I can do here.

BrandonWilliamsCS commented:
Pinging @janicduplessis just to make sure my update didn't get lost in the shuffle of the holidays.

janicduplessis commented:
Thanks @BrandonWilliamsCS, will merge after holidays!

BrandonWilliamsCS commented:
@janicduplessis I don't mean to be a pest, but is this still slated for merge?

janicduplessis commented:
Thanks for reminding me, yes!

@facebook-github-bot shipit

colinux commented:
This PR seems to not have been merged.

I'm facing the same issue on RN 64, which should have been fixed by this PR, and I don't see this code in master. Or did I miss something ?

hikaaam commented:
this issue still happen react-native v.0.63, keyboardType visible-password make my password visible
